### PR TITLE
BUG: Dealloc cached buffer info

### LINKED ISSUE #12122: BUG memory leak using ProcessPoolExecutor
Hello,
A recent [bug report](https://github.com/joblib/joblib/issues/781) was filed in [joblib](https://github.com/joblib/joblib), alerting us about a memory leak when calling several times a `Parallel` instance on a numpy.ndarray iterator.

While trying to recreate a minimal example, it turns out that the bug still exists without `joblib`,  when using only numpy and a `concurent.futures.ProcessPoolExecutor`.

Running the following bit of code on a fresh python 3.6 virtual environment with
```
numpy==1.15.2
psutil==5.4.7
``` 
```python
import numpy as np
import psutil
import gc

from concurrent.futures import ProcessPoolExecutor


def func(*args, **kw):
    pass


if __name__ == "__main__":
    n = 100000
    x = np.arange(n)
    print("x size: %0.3f MB" % (x.nbytes / 1e6))

    executor = ProcessPoolExecutor(2)

    for i in range(10):
        gc.collect()
        print("Iteration %d: %0.3f MB" %
              (i, psutil.Process().memory_info().rss / 1e6))

        for r in executor.map(func, [x[i] for i in range(n)], chunksize=100):
            pass

        # The following for-loop does not cause the memory leak
        # for r in executor.map(
        #         func, [x[i:i + 1] for i in range(n)], chunksize=100):
        #     pass
```

gives out:

```
x size: 0.800 MB
Iteration 0: 30.200 MB
Iteration 1: 61.817 MB
Iteration 2: 92.025 MB
Iteration 3: 116.171 MB
Iteration 4: 151.134 MB
Iteration 5: 174.965 MB
Iteration 6: 199.463 MB
Iteration 7: 223.617 MB
Iteration 8: 268.849 MB
Iteration 9: 293.966 MB
```

The problem pretty much  disappears in *all* of these cases:
* Running the commented for-loop instead 
* Using `concurent.futures.ThreadPoolExecutor` instead of `concurrent.futures.ProcessPoolExecutor`
* Calling `[pickle.dumps(xi) for xi in x]`
* Iterating on a `range` or a `list` instead of x by calling for example `x = x.tolist()`
* using `x = np.arange(n).reshape(-1, 1)` instead of `x = np.arange(n)`
* Populating and emptying a `multiprocessing.Queue` with the elements of `x` (by running this chunk of code instead of the `for r in executor.map(...)` loop):

```python
        from multiprocessing import Queue
        q = Queue()
        [q.put(xi) for xi in x]
        [q.get() for _ in range(n)]
        q.close()
```

Let me know if you want more detailed output about any of these cases.

We are kind of out of ideas with @ogrisel, so we are hoping to have some insight from more experienced numpy developers :)
Pierre

### PR DESCRIPTION
**Pull request by mattip**
Fixes #12122.

577dbbd, added in 2009 for numpy 1.5 to fix TRAC issue 1312 worked around a deficiency in that "`PyArg_ParseTuple("s#", ...)` refuses to return a buffer if ``bf_releasebuffer`` is present". That changeset never released a buffer_info_cache entry for a scalar. This fixes that problem. Note also cpython [issue 24009](https://bugs.python.org/issue24009) where they discuss deprecating [`"s#"`](https://docs.python.org/2.0/ext/parseTuple.html) parsing.

It is difficult to test this fix since the cache keys are id(value) and the cache is `static` in `buffer.c`. We should revisit the need for the cache, it seems like there should be a better way.

Edit: note that the issue number predates github, formatting

### CODE REVIEW

**Comment by eric-wieser** [comment]
Trac 1312 is #1910

**Comment by pierreglaser** [comment]
Hi All, I gave an update on this in #12122 :)

File: numpy/core/src/multiarray/buffer.h [hunk 1]
```diff
@@ -4,7 +4,7 @@
 extern NPY_NO_EXPORT PyBufferProcs array_as_buffer;
 
 NPY_NO_EXPORT void
-_array_dealloc_buffer_info(PyArrayObject *self);
+_array_dealloc_buffer_info(PyObject *self);
```
Comment (pierreglaser) [inline numpy/core/src/multiarray/buffer.h]: Should it still be called `_array_dealloc_buffer_info` as now it is used to deallocate `buffer_info` from other objects?

**Comment by mattip** [comment]
a749c1834 actually was for descriptors, not scalars. c0e515b78 is the fix for scalars.

In changeset 7258368 I also added preemptive calls to `_array_dealloc_buffer_info` as soon as possible, i.e. right after `PyBuffer_Release`

In an ideal world, `array_dealloc_buffer_info` would be part of ndarray's `bf_releasebuffer`, but we cannot use that without breaking various `PyArgParse` routines that use `s#`

**Comment by pierreglaser** [comment]
@mattip with the new batch of commits you just pushed, memory usage is back to normal on my test cases so, great! :) 
For the record, I think the key fix was adding `array_dealloc_buffer_info` during `gentype_dealloc`. Actually, with that line of code only, the leak is contained.

**Comment by charris** [comment]
I everyone happy with this?

**Pull request merged**